monitor/services/src/main/java/org/datacleaner/monitor/server/SchedulingServiceImpl.java

```diff
@@ -131,16 +137,80 @@ private HotFolderAlterationListener(final JobIdentifier job, final ScheduleDefin
         public void onFileCreate(final File file) {
             logger.info("file {} created in hot folder, triggering execution of job {}.", file.getName(),
                     job.getName());
-
-            triggerJobExecution();
+            executeJobWithFile(file);
         }
 
         @Override
         public void onFileChange(final File file) {
             logger.info("file {} changed in hot folder, triggering execution of job {}.", file.getName(),
                     job.getName());
+            executeJobWithFile(file);
+        }
+
+        private void executeJobWithFile(final File file) {
+            if (!file.exists()) {
+                logger.warn("Data file '{}' triggering execution of the job '{}' does not exist.",
+                        file.getAbsolutePath(), job.getName());
+                return;
+            }
+
+            final String dataStoreName = getDataStoreName();
 
-            triggerJobExecution();
+            if (dataStoreName != null) {
+                try {
+                    waitForCompleteFile(file);
+                } catch (final IncompleteFileException e) {
+                    logger.error("Hot folder job execution failed because the file '{}' is incomplete of unavailable. ",
+                            file.getAbsolutePath());
+                    return;
+                }
+
+                final Map<String, String> propertiesMap = new HashMap<>();
+                propertiesMap.put("datastoreCatalog." + getDataStoreName() + ".filename", file.getAbsolutePath());
+                triggerExecution(tenant, job, propertiesMap, TriggerType.HOTFOLDER);
+            } else {
+                triggerJobExecution();
+            }
+        }
+
+        /**
+         * File event can be triggered before the complete file content is written (copying in progress).
+         * This method tries to check that possibility and wait for the complete file to be ready.
+         * Unfortunately it is not file/operating system independent.
+         */
+        private void waitForCompleteFile(final File file) throws IncompleteFileException {
+            final WaitForCompleteFileStrategy waitStrategy;
+            final String osName = System.getProperty("os.name").toLowerCase();
+
+            if (osName.contains("windows")) {
+                logger.info("Using WindowsWaitForCompleteFileStrategy. ");
+                waitStrategy = new WindowsWaitForCompleteFileStrategy();
+            } else if (osName.contains("linux")) {
+                logger.info("Using LinuxWaitForCompleteFileStrategy. ");
+                waitStrategy = new LinuxWaitForCompleteFileStrategy();
+            } else {
+                logger.info("Using GeneralWaitForCompleteFileStrategy. ");
+                waitStrategy = new GeneralWaitForCompleteFileStrategy();
+            }
+
+            waitStrategy.waitForComplete(file);
```

Comment:
Still not entirely what I was looking for, better though.

But at least use one of the existing thread pools.